File: 程序员面试金典第6版/面试题 16.04. Tic-Tac-Toe LCCI.cpp
// C++ 可以这么写 因为C/C++ 里面 非0数字都表示true 仅有0表示false
class Solution {
public:
    string tictactoe(vector<string>& board) {
        int n = board.size();
        bool full = true;
        bool diag1XWin = false, diag1OWin = false;
        bool diag2XWin = false, diag2OWin = false;

        for(int i=0; i<n; i++) {
            bool rowXWin = false, rowOWin = false;
            bool colXWin = false, colOWin = false;
            for(int j=0; j<n; j++) {
                if(full && board[i][j] == ' ') {
                    full = false;
                }
                if(!rowXWin) rowXWin = rowXWin || ('X' ^ board[i][j]);
                if(!rowOWin) rowOWin = rowOWin || ('O' ^ board[i][j]);
                if(!colXWin) colXWin = colXWin || ('X' ^ board[j][i]);
                if(!colOWin) colOWin = colOWin || ('O' ^ board[j][i]);
            }

            if(rowXWin == false || colXWin == false) return "X";
            if(!rowOWin || !colOWin) return "O";
            // diagnonal
            diag1XWin = diag1XWin || ('X' ^ board[i][i]);
            diag1OWin = diag1OWin || ('O' ^ board[i][i]);
            diag2XWin = diag2XWin || ('X' ^ board[i][n-i-1]);
            diag2OWin = diag2OWin || ('O' ^ board[i][n-i-1]);
        }

        if(!diag1XWin || !diag2XWin) return "X";
        if(!diag1OWin || !diag2OWin) return "O";

        if(full) {
            return "Draw";
        } else {
            return "Pending";
        }
    }
};

File: 程序员面试金典第6版/面试题 16.04. Tic-Tac-Toe LCCI.java
 class Solution {
    public String tictactoe(String[] board) {
        int n = board.length;
        int rowSum = 0, colSum = 0, diag1Sum = 0, diag2Sum = 0;
        boolean flag = false;

        for(int i=0; i<n; i++) {
            rowSum = 0;
            colSum = 0;
            for(int j=0; j<n; j++) {
                rowSum += board[i].charAt(j);
                colSum += board[j].charAt(i);
                if(!flag && board[i].charAt(j) == ' ') {
                    flag = true;
                }
            }
            if(rowSum == 'X'*n) return "X";
            if(rowSum == 'O'*n) return "O";
            if(colSum == 'X'*n) return "X";
            if(colSum == 'O'*n) return "O";
            diag1Sum += board[i].charAt(i);
            diag2Sum += board[i].charAt(n-i-1);
        }
        if(diag1Sum == 'X'*n) return "X";
        if(diag1Sum == 'O'*n) return "O";
        if(diag2Sum == 'X'*n) return "X";
        if(diag2Sum == 'O'*n) return "O";

        if(flag) {
            return "Pending";
        } else {
            return "Draw";
        }
    }
}